shopinvader/services/abstract_mail.py

```diff
@@ -0,0 +1,103 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 ACSONE SA/NV (<http://acsone.eu>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from odoo.addons.component.core import AbstractComponent
+
+
+class AbstractMailService(AbstractComponent):
+    _inherit = 'base.shopinvader.service'
+    _name = 'shopinvader.abstract.mail.service'
+
+    def ask_email(self, _id):
+        """
+        Ask to receive the record ID by email
+        :param _id: int
+        :return:
+        """
+        self._ask_email(_id)
+        return {}
+
+    def _validator_ask_email(self):
+        return {}
+
+    def _is_logged(self):
+        """
+        Check if the current partner is a real partner (not the anonymous one
+        and not empty)
+        :return: bool
+        """
+        logged = False
+        if hasattr(self, 'partner') and self.partner:
+            if self.partner != self.shopinvader_backend.anonymous_partner_id:
+                logged = True
+        return logged
+
+    def _get_email_notification_type(self, record):
+        """
+        Based on the given record, get the notification type.
+        By default, it's build like this:
+        <_usage>_send_email
+        :param record: target record
+        :return: str
+        """
+        if hasattr(self, '_usage') and self._usage:
+            return '%s_send_email' % self._usage
+        return ''
+
+    def _load_target_email(self, record_id):
+        """
+        Load the record (based on expose model)
+        :param record_id: int
+        :return: record or None
+        """
+        if not getattr(self, '_expose_model', False):
+            return None
+        return self.env[self._expose_model].browse(record_id)
+
+    def _allow_email_notification(self, partner, record, notif_type):
+        """
+        Based on given record, partner and notif_type, check if the partner
+        is allowed to launch this kind of notification
+        :param partner: res.partner
+        :param record: target record
+        :param notif_type: str
+        :return: bool
+        """
+        if not notif_type:
+            return False
+        if record and hasattr(record, 'partner_id') and record.partner_id:
```

Comment:
@acsonefho You must add a method into this abtract to get the recipient from the record..

```
def _get_email_recipient(self, record, notif_type):
   # default implementation 
    return record.partner_id
```

If the recipient is not the partner_id field, the implementation service must override this method